libraries/state_history/trace_converter.cpp

```diff
@@ -30,25 +38,262 @@ void trace_converter::add_transaction(const transaction_trace_ptr& trace, const
    std::vector<augmented_transaction_trace> traces;
-   if (onblock_trace)
-      traces.push_back(*onblock_trace);
+   if (converter.onblock_trace)
+      traces.push_back(*converter.onblock_trace);
    for (auto& r : block_state->block->transactions) {
       transaction_id_type id;
       if (r.trx.contains<transaction_id_type>())
          id = r.trx.get<transaction_id_type>();
       else
          id = r.trx.get<packed_transaction>().id();
-      auto it = cached_traces.find(id);
-      EOS_ASSERT(it != cached_traces.end() && it->second.trace->receipt, plugin_exception,
+      auto it = converter.cached_traces.find(id);
+      EOS_ASSERT(it != converter.cached_traces.end() && it->second.trace->receipt, state_history_exception,
                  "missing trace for transaction ${id}", ("id", id));
       traces.push_back(it->second);
    }
-   cached_traces.clear();
-   onblock_trace.reset();
+   converter.cached_traces.clear();
+   converter.onblock_trace.reset();
+   return traces;
+}
+
+template <typename Lambda>
+void for_each_packed_transaction(const eosio::state_history::augmented_transaction_trace& obj, const Lambda& lambda) {
+   auto& trace = *obj.trace;
+   if (trace.failed_dtrx_trace) {
+      for_each_packed_transaction(
+          eosio::state_history::augmented_transaction_trace{trace.failed_dtrx_trace, obj.packed_trx}, lambda);
+   }
+   bool include_packed_trx = obj.packed_trx && !trace.failed_dtrx_trace;
+   if (include_packed_trx) {
+      lambda(*obj.packed_trx);
+   }
+}
+
+template <typename Lambda>
+void for_each_packed_transaction(const std::vector<eosio::state_history::augmented_transaction_trace>& traces,
+                                 const Lambda&                                                         lambda) {
+   for (const auto& v : traces) {
+      for_each_packed_transaction(v, lambda);
+   }
+}
+
+void pack(fc::datastream<char*>& ds, const prunable_data_type&  data, compression_type compression) {
+   if (compression == compression_type::none) {
+      fc::raw::pack(ds, data);
+   }
+   else if (compression == compression_type::zlib) {
+      bytes uncompressed = fc::raw::pack(data);
+      fc::raw::pack(ds, zlib_compress_bytes(uncompressed));
+   }
+}
+
+size_t pack_prunable(bytes& buffer, const packed_transaction& trx,
+                  compression_type compression) {
+
+   const auto& data = trx.get_prunable_data();
+
+   int max_size = data.maximum_pruned_pack_size(compression);
+   uint64_t start_buffer_position = buffer.size();
+   buffer.resize(buffer.size() + max_size);
+   fc::datastream<char*> ds(buffer.data() + start_buffer_position, max_size);
+
+   if (data.prunable_data.contains<prunable_data_type::full_legacy>()) {
+      // convert full_legacy to full before serialization
+      const auto& full_legacy  = data.prunable_data.get<prunable_data_type::full_legacy>();
+      prunable_data_type pd    = {prunable_data_type::full{full_legacy.signatures, full_legacy.context_free_segments}};
```

Comment:
Instead of paying for a copy, how about packing it directly. 
Add: 
`static_assert(fc::reflector<prunable_data_type::full>::total_member_count == 2);`
Then just pack the `signatures` & `context_free_segments`.
This could be added to `pack(fc::datastream<char*>& ds, const prunable_data_type&  data, compression_type compression`

Reply:
Since we are just storing the signature and context_free_segments always in the same form for full and full_legacy. Can't we just store them in the as we like making sure to leave enough room for a digest? So can't we just always compress? Doesn't matter what the packed_transaction has.